File: sdk/dart/api/lib/module/zlg_device_command.dart
import 'package:meta/meta.dart';

import '../base/client_error_code.dart';
import '../base/zlg_base_module.dart';
import '../framework/zlg_http.dart';
import '../util/util.dart';

class ZLGDeviceCommand extends ZLGBaseModule {
  static ZLGDeviceCommand _instance;

  factory ZLGDeviceCommand() {
    if (_instance == null) {
      _instance = ZLGDeviceCommand._internal();
    }
    return _instance;
  }

  ZLGDeviceCommand._internal();

  Future getDeviceCommands(
      {@required String deviceType, @required String deviceId}) {
    assert(deviceType != null && deviceId != null);
    return fetch(
        url: '/devices/$deviceId/commands',
        method: Methods.get,
        params: {'devtype': deviceType}).catchError((err) {
      if (err is DioError) {
        var errorMap = <int, int>{404: ClientErrorCode.ZLG_DEVICE_NOT_EXIST};
        handleError(errorMap, err.response);
      }
      throw err;
    });
  }

  Future sendCommandToDevice(
      {@required String deviceType,
      @required String deviceId,
      num timeout,
      @required String cmdType,
      @required String cmd,
      @required Map<String, dynamic> args}) {
    assert(
        deviceType != null && deviceId != null && cmd != null && args != null);
    assert(cmdType == 'confirmed' || cmdType == 'unconfirmed');
    return fetch(
            url: '/devices/$deviceId/commands/$cmd',
            method: Methods.put,
            params: Util.deleteEmptyMapEntity({
              'devtype': deviceType,
              'timeout': timeout,
              'cmdtype': cmdType
            }),
            data: Util.deleteEmptyMapEntity(args))
        .catchError((err) {
      if (err is DioError) {
        var errorMap = <int, int>{404: ClientErrorCode.ZLG_COMMAND_NOT_EXIST};
        handleError(errorMap, err.response);
      }
      throw err;
    });
  }

  Future sendRawCommandToDevice(
      {@required String deviceType,
      @required String deviceId,
      @required String dataType,
      @required String raw}) {
    assert(deviceType != null &&
        deviceId != null &&
        dataType != null &&
        raw != null);
    assert(dataType == 'hex' || dataType == 'ascii' || dataType == 'base64');
    return fetch(
        url: '/devices/$deviceId/command/raw',
        method: Methods.put,
        params: Util.deleteEmptyMapEntity(
            {'devtype': deviceType, 'datatype': dataType}),
        data: Util.deleteEmptyMapEntity({'raw': raw})).catchError((err) {
      if (err is DioError) {
        var errorMap = <int, int>{};
        handleError(errorMap, err.response);
      }
      throw err;
    });
  }

  Future sendStatusCommandToDevice(
      {@required String deviceType,
      @required String deviceId,
      int timeout,
      @required String cmdType,
      @required Map<String, dynamic> args}) {
    assert(deviceType != null &&
        deviceId != null &&
        cmdType != null &&
        args != null);
    print('sendStatusCommandToDevice args: $args');
    return fetch(
            url: '/devices/$deviceId/commands/set_config',
            method: Methods.put,
            params: Util.deleteEmptyMapEntity({
              'devtype': deviceType,
              'timeout': timeout,
              'cmdtype': cmdType
            }),
            data: Util.deleteEmptyMapEntity(args))
        .catchError((err) {
      if (err is DioError) {
        var errorMap = <int, int>{
          404: ClientErrorCode.ZLG_COMMAND_NOT_EXIST,
          408: ClientErrorCode.ZLG_SEND_COMMAND_NO_RSP
        };
        handleError(errorMap, err.response);
      }
      throw err;
    });
  }
}

File: sdk/js/api/http/httpModules/ZLGDeviceCommand.js
import ZLGBaseModule from '../httpBase/ZLGBaseModule'
import ZLGErrors from '../httpBase/ZLGHttpClientCode'

/**
 * 设备命令
 */
class ZLGDeviceCommand extends ZLGBaseModule {
  /**
   * 获取设备支持的命令
   * @param {Object} params
   * @param {string} params.devtype - 设备类型
   * @param {string} params.devid - 设备ID
   * @return {Promise<DeviceCommnd[], Error>} - response: [{@link DeviceCommnd}]
   */
  static getDeviceCommands ({ devtype, devid }) {
    return this.fetch({
      url: `/devices/${devid}/commands`,
      method: 'get',
      params: { devtype }
    }).catch(err => {
      let errorMap = {
        404: ZLGErrors.ZLG_DEVICE_NOT_EXIST
      }
      return this.handleError(errorMap, err)
    })
  }
  /**
   * 向指定设备发送命令
   * @param {Object} params
   * @param {string} params.devtype - 设备类型
   * @param {string} params.devid - 设备ID
   * @param {number} params.timeout - 命令超时时间(ms)
   * @param {number} params.cmdtype - 命令类型
   * @param {number} params.cmd - 命令的名称
   * @param {number} params.args - 命令的参数
   * @return {Promise<Response, Error>} -
   */
  static sendCommandToDevice ({ devtype, devid, timeout, cmdtype, cmd, args }) {
    return this.fetch({
      url: `/devices/${devid}/commands/${cmd}`,
      method: 'put',
      params: { devtype, timeout, cmdtype },
      data: args
    }).catch(err => {
      let errorMap = {
        404: ZLGErrors.ZLG_COMMAND_NOT_EXIST
      }
      return this.handleError(errorMap, err)
    })
  }
  /**
   * 向指定设备发送配置设备状态信息命令
   * @param {Object} params
   * @param {string} params.devtype - 设备类型
   * @param {string} params.devid - 设备ID
   * @param {number} params.timeout - 命令超时时间(ms)
   * @param {number} params.cmdtype - 命令类型
   * @param {number} params.args - 命令的参数
   * @return {Promise<Response, Error>} -
   */
  static sendStatusCommandToDevice ({ devtype, devid, timeout, cmdtype, args }) {
    return this.fetch({
      url: `/devices/${devid}/commands/set_config`,
      method: 'put',
      params: { devtype, timeout, cmdtype },
      data: args
    }).catch(err => {
      let errorMap = {
        404: ZLGErrors.ZLG_COMMAND_NOT_EXIST
      }
      return this.handleError(errorMap, err)
    })
  }
  /**
   * 发送原始数据到指定设备
   * @param {Object} params
   * @param {string} params.devtype - 设备类型
   * @param {string} params.devid - 设备ID
   * @param {string} params.info -
   * @param {string} params.info.raw - 原始数据(base64编码字符串)
   * @return {Promise<Response, Error>} -
   */
  static sendRawCommandToDevice ({ devtype, devid, info }) {
    return this.fetch({
      url: `/devices/${devid}/commands/raw`,
      method: 'put',
      data: info,
      params: { devtype }
    }).catch(err => {
      let errorMap = {}
      return this.handleError(errorMap, err)
    })
  }
}

export default ZLGDeviceCommand
